iib/workers/tasks/opm_operations.py

```diff
@@ -0,0 +1,169 @@
+import logging
+import os
+import socket
+import subprocess
+import time
+
+from retry import retry
+
+from iib.exceptions import AddressAlreadyInUse, IIBError
+from iib.workers.config import get_worker_config
+from iib.workers.tasks.fbc_utils import is_image_fbc, get_catalog_dir
+
+log = logging.getLogger(__name__)
+
+
+def _get_free_port(port_start, port_end):
+    """
+    Return free port that is safe to use for opm command.
+
+    :param int port_start: port from which we should start trying to connect to
+    :param int port_end: port on which we should stop trying
+    """
+    for port in range(port_start, port_end):
+        sock = socket.socket()
+        try:
+            sock.bind(('', port))
+            sock.close()
+            return port
+
+        except OSError:
+            log.info('Port %d is in use, trying another.', port)
+            port += 1
+        sock.close()
+
+    err_msg = f'No free port has been found after {port_end - port_start} attempts.'
+    log.error(err_msg)
+    raise IIBError(err_msg)
+
+
+def opm_serve_from_index(base_dir, from_index):
+    """
+    Locally start OPM registry service, which can be communicated with using gRPC queries.
+
+    Due to IIB's paralellism, the service can run multiple times, which could lead to port
+    binding conflicts. Resolution of port conflicts is handled in this function as well.
+
+    :param str base_dir: base directory to create temporary files in.
+    :param str from_index: index image to inspect.
+    :return: tuple containing port number of the running service and the running Popen object.
+    :rtype: (int, Popen)
+    :raises IIBError: if all tried ports are in use, or the command failed for another reason.
+    """
+    from iib.workers.tasks.build import _get_index_database
+
+    if not is_image_fbc(from_index):
+        db_path = _get_index_database(from_index, base_dir)
+        return opm_registry_serve(db_path)
+
+    catalog_dir = get_catalog_dir(from_index, base_dir)
+    return opm_serve(catalog_dir)
+
+
+def opm_serve(catalog_dir):
```

Comment:
This function seems to be almost the same with opm_registry_serve. Can we distinguish by name if this is used for  FBC?

Reply:
ad naming: I was thinking about it however I went with this because it directly reflects what is called `opm serve` and `opm registry serve`. This is how it is distinguish in `opm`, should we do it differently? 
Is not doc-string description of the function sufficient?

Thanks

Reply:
sick day?

Reply:
Yes. But we shod deal with this PR asap :) 
It is blocking future work.

Reply:
The functions `opm_serve` and `opm_registry_serve` are indeed quite similar, the only differences are:

1. Parameter `catalog_dir` vs `db_path`
2. Inside function: `cmd` and `cwd` content differs

And that's it, the rest of the functions are simply copy+paste

One suggestion to avoid this would be creating something like that (if you think that's useful, otherwise I'm open to suggestions):

```{python}
def opm_serve(base_dir, is_fbc=True):
    [...]
    if is_fbc:
        cmd = ['opm', 'serve', catalog_dir, '-p', str(port), '-t', '/dev/null']
        cwd = os.path.abspath(os.path.join(catalog_dir, os.path.pardir))
    else:
        cmd = ['opm', 'registry', 'serve', '-p', str(port), '-d', db_path, '-t', '/dev/null']
        cwd = os.path.dirname(db_path)
    [...]

```

Reply:
I had it before however Yash suggested to split it to two because they are used to different things.
What you are suggesting leads to having different inputs for one function and I do like this separation better.

Reply:
I updated the code and simplified it a bit, I still prefer to not mix those functions but the duplicate code was separated.

Reply:
@JAVGan please check.. I am ok with the changes

Reply:
Sure, I got the point, I'm ok with it too.

Reply:
Closing